### PR TITLE
Issue #3545: Remove LocalizedMessages class

### PR DESCRIPTION
**Pull request by MEZk**
Issue #3545

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 1]
```diff
@@ -37,7 +38,7 @@
     implements FileSetCheck {
 
     /** Collects the error messages. */
-    private final LocalizedMessages messageCollector = new LocalizedMessages();
+    private final SortedSet<LocalizedMessage> errorMessages = new TreeSet<>();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: why rename this to `errorMessages` (specifically `error`) when you want it to be `violation` at #4465 ?

Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Renaming will be done in #4465 when we introduce CheckstyleViolation as a part of our model.

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: So we are going to rename once here and then rename again in another issue?
Why don't we just keep the same name here then?

Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: > Why don't we just keep the same name here then and rename only once in other issue?

Ok, done.

**Comment by MEZk** [comment]
@rnveach @romani 
Looks like we have the design problem. Look at the IDEA's inspection violation for the method ```AbstractCheck#setMessages```.

> Assignment to Collection or array field from parameter
Reports any attempt to assign an array or Collection field from a method parameter. Since the array or Collection may have its contents modified by the calling method, this construct may result in an object having its state modified unexpectedly. While occasionally useful for performance reasons, this construct is inherently bug-prone.

**Comment by rnveach** [comment]
> We cannot just create a copy of the collection as the reference should be stored in the global context.

@MEZk This is the same issue we ran into at #4306. You must create new list from parameter, you can't do direct assignment otherwise the list can be modified outside the class.

**Comment by MEZk** [comment]
@rnveach 

> You must create new list from parameter,

The problem is in TreeWalker design. We store the reference to the message collector [here](https://github.com/MEZk/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L142) then we invoke [contextualize](https://github.com/MEZk/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L161-L161) and access the stored reference [here](https://github.com/MEZk/checkstyle/blob/4cef3f91f575f45ea18fa3568ec44c0beef6a1fe/src/main/java/com/puppycrawl/tools/checkstyle/api/AutomaticBean.java#L234-L234). If we create the new set from the parameter, we will [add new localized messages](https://github.com/MEZk/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java#L239-L239) to the copy, but not to the set which is stored in the TreeWalker context.

**Comment by MEZk** [comment]
@romani 
Please, help us to resolve the problem, share your opinion on the design.

**Comment by romani** [comment]
We need one more step and remove `public final void setMessages(LocalizedMessages messages) {` from Checks, each Check should have it's own collection of messages, parent module grab it from Check after Check finished

**Comment by MEZk** [comment]
@romani 
1) Should TreeWalker grab the collection of messages from each check in notifyEnd and add them into its own collection of messages?
2) Should we change ```AbstractFileSetCheck#processFiltered``` in order to return a sorted set of localized messages?

**Comment by romani** [comment]
@MEZk ,

> Should TreeWalker grab the collection of messages from each check in notifyEnd and add them into its own collection of messages?

yes.
There should be no global collections.
It should ok to use reflection (only for now) to initialize Check with parameters and let Check go independedntly and be able to get from it list of violations at the end.

> Should we change AbstractFileSetCheck#processFiltered in order to return a sorted set of localized messages?

probably yes, please change and we will see how it looks like.

**Comment by romani** [comment]
Please also rebase on latest master to test travis fix.

**Comment by MEZk** [comment]
> Please also rebase on latest master to test travis fix.

Done.

> Should we change AbstractFileSetCheck#processFiltered in order to return a sorted set of localized messages?

This will not work as all inheritors of AbstractFileSetCheck use `log` method to add a new localized message into the sorted set of localized messages held by AbstractFileSetCheck. If we try to get the set of messages from AbstractCheck in 'notifyEnd' method of TreeWalker, and add them into the sorted set of localized messages (the field of TreeWalker), then we will need to add the set from TreeWalker into the set from AbstractFileSetCheck. Thus, there will be a design problem: all classes which extend AbstractFileSetCheck except for TreeWalker will modify the set of localized messages from AbstractFileSetCheck through 'log' method, but TreeWalker will have the different way of doing the same thing. Even though TreeWalker is a complex composite module, I do not believe that it should have such privileges.

**Comment by MEZk** [comment]
Rebased onto the latest master.
Fixed TeamCity violation.

**Comment by romani** [comment]
checkstyle 8.0 was released, please rebase on latest master. 
after each release it is mandatory to rebase to avoid CIs failures.

**Comment by MEZk** [comment]
@romani 
Done.

**Comment by romani** [comment]
please rebase one more time, I just fixed master build failure for Travis and Appveyor

**Comment by MEZk** [comment]
@romani 
Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java [hunk 2]
```diff
@@ -352,6 +351,8 @@ private void notifyEnd(DetailAST rootAST, AstState astState) {
 
         for (AbstractCheck check : checks) {
             check.finishTree(rootAST);
+            addMessages(check.getMessages());
+            check.clearMessages();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java]: @romani If any check has exception in `finishTree` or `visitToken`, messages will never be cleared and appear in the next file run. We saw something like this in sevntu. https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/589 .
IMO, we should clear messages right before `beginTree`.
We are already doing this in `AbstractFileSetCheck` in this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 3]
```diff
@@ -134,14 +135,11 @@ public final void setFileExtensions(String... extensions) {
...
-     * Subclasses can use the collector to find out the violation
-     * messages to fire via the message dispatcher.
-     *
-     * @return the collector for localized messages.
+     * Adds the sorted set of {@link LocalizedMessage} to the message collector.
+     * @param messages the sorted set of {@link LocalizedMessage}.
      */
-    protected final LocalizedMessages getMessageCollector() {
-        return messageCollector;
+    public final void addMessages(SortedSet<LocalizedMessage> messages) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Why `public`? I don't see any external calls in this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java (hunk 2 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java]: Make sense. Moved ``` check.clearMessages()``` to ```beginTree```.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java (hunk 3 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: changed to protected

**Comment by romani** [comment]
I did review one more time , nothing spotted.
As soon as Shippable is green, we could merge

**Pull request merged**